daemon/mgr/spec_namespace.go

```diff
@@ -2,39 +2,148 @@ package mgr
+func isNone(mode string) bool {
+	return mode == "none"
+}
+
+// isHost indicates whether the container shares the host's corresponding namespace.
+func isHost(mode string) bool {
+	return mode == "host"
+}
+
+// isShareable indicates whether the containers namespace can be shared with another container.
+func isShareable(mode string) bool {
+	return mode == "shareable"
+}
+
+// isContainer indicates whether the container uses another container's corresponding namespace.
+func isContainer(mode string) bool {
+	parts := strings.SplitN(mode, ":", 2)
+	return len(parts) > 1 && parts[0] == "container"
+}
+
+// isPrivate indicates whether the container uses its own namespace.
+func isPrivate(ns specs.LinuxNamespaceType, mode string) bool {
+	switch ns {
+	case specs.IPCNamespace:
+		return mode == "private"
+	case specs.NetworkNamespace, specs.PIDNamespace:
+		return !(isHost(mode) || isContainer(mode))
+	case specs.UserNamespace, specs.UTSNamespace:
+		return !(isHost(mode))
+	}
+	return false
+}
+
+// connectedContainer is the id or name of the container whose namespace this container share with.
+func connectedContainer(mode string) string {
+	parts := strings.SplitN(mode, ":", 2)
+	if len(parts) == 2 {
+		return parts[1]
+	}
+	return ""
+}
+
+func getIpcContainer(ctx context.Context, mgr ContainerMgr, id string) (*ContainerMeta, error) {
+	// Check whether the container exists.
+	c, err := mgr.Get(ctx, id)
+	if err != nil {
+		return nil, fmt.Errorf("can't join IPC namespace of container %q: %v", id, err)
+	}
+
+	// TODO: check whether the container is running and not restarting.
+
+	// TODO: check whether the container's ipc namespace is shareable.
+
+	return c, nil
+}
+
+func getPidContainer(ctx context.Context, mgr ContainerMgr, id string) (*ContainerMeta, error) {
+	// Check the container exists.
+	c, err := mgr.Get(ctx, id)
+	if err != nil {
+		return nil, fmt.Errorf("can't join PID namespace of %q: %v", id, err)
+	}
+
+	// TODO: check whether the container is running and not restarting.
+
+	return c, nil
+}
+
 // TODO
-func setupUserNamespace(ctx context.Context, meta *ContainerMeta, s *specs.Spec) error {
+func setupUserNamespace(ctx context.Context, mgr ContainerMgr, meta *ContainerMeta, s *specs.Spec) error {
 	return nil
 }
 
 // TODO
-func setupNetworkNamespace(ctx context.Context, meta *ContainerMeta, s *specs.Spec) error {
+func setupNetworkNamespace(ctx context.Context, mgr ContainerMgr, meta *ContainerMeta, s *specs.Spec) error {
 	return nil
 }
 
-func setupIpcNamespace(ctx context.Context, meta *ContainerMeta, s *specs.Spec) error {
-	ns := specs.LinuxNamespace{Type: specs.IPCNamespace}
-	setNamespace(s, ns)
+func setupIpcNamespace(ctx context.Context, mgr ContainerMgr, meta *ContainerMeta, s *specs.Spec) error {
+	ipcMode := meta.HostConfig.IpcMode
+	switch {
+	case isContainer(ipcMode):
+		ns := specs.LinuxNamespace{Type: specs.IPCNamespace}
+		c, err := getIpcContainer(ctx, mgr, connectedContainer(ipcMode))
```

Comment:
If the container from connectedContainer(ipcMode) is same as current container? Could We Add a Judgement to do it.

Reply:
Hmmm, actually we should only get running container from getIpcContainer, and the current container couldn't be running.

So it is safe here 😄